cpu/sam0_common/sam0_sdhc/sdhc.c

```diff
@@ -0,0 +1,1062 @@
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @ingroup     cpu_sam0_common_sdhc
+ * @{
+ *
+ * @file
+ * @brief       Driver for the SD Host Controller
+ *
+ * @author      alkgrove <bob@alkgrove.com>
+ * @author      Benjamin Valentin <benjamin.valentin@ml-pa.com>
+ *
+ * @}
+ */
+
+#include <errno.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <stdio.h>
+#include "periph_cpu.h"
+#include "periph/pm.h"
+#include "vendor/sd_mmc_protocol.h"
+#include "sdhc.h"
+#include "time_units.h"
+#include "ztimer.h"
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+#ifndef SDHC_CLOCK
+#define SDHC_CLOCK          SAM0_GCLK_MAIN
+#endif
+
+#ifndef SDHC_CLOCK_SLOW
+#define SDHC_CLOCK_SLOW     SAM0_GCLK_TIMER
+#endif
+
+/**
+ * @brief   The board can overwrite this if only a single SDHC instance is used
+ *          to save 80 Bytes of ROM.
+ */
+#ifndef SDHC_DEV
+#ifdef SDHC1
+#define SDHC_DEV            state->dev
+#else
+#define SDHC_DEV            SDHC0
+#endif
+#endif
+
+/**
+ * @brief   Monitor card insertion and removal
+ */
+#define NISTR_CARD_DETECT   (SDHC_NISTR_CREM | SDHC_NISTR_CINS)
+
+static sdhc_state_t *isr_ctx_0;
+static sdhc_state_t *isr_ctx_1;
+
+static void _set_speed(sdhc_state_t *state, uint32_t fsdhc);
+static void _set_hc(sdhc_state_t *state);
+static bool _test_voltage(sdhc_state_t *state, bool f8);
+static bool _test_capacity(sdhc_state_t *state);
+static bool _test_version(sdhc_state_t *state);
+static bool _wait_not_busy(sdhc_state_t *state);
+static bool _test_bus_width(sdhc_state_t *state);
+static bool _test_high_speed(sdhc_state_t *state);
+static bool _init_transfe(sdhc_state_t *state, uint32_t cmd, uint32_t arg, uint16_t block_size,
+                          uint16_t num_blocks);
+static bool sdio_test_type(sdhc_state_t *state);
+
+/** SD/MMC transfer rate unit codes (10K) list */
+static const uint32_t transfer_units[] = { 10, 100, 1000, 10000, 0, 0, 0, 0 };
+/** SD transfer multiplier factor codes (1/10) list */
+static const uint8_t transfer_multiplier[16] =
+{ 0, 10, 12, 13, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 70, 80 };
+
+static bool _card_detect(sdhc_state_t *state)
+{
+    return state->dev->PSR.bit.CARDINS;
+}
+
+static bool _check_mask(uint32_t val, uint32_t mask)
+{
+    return (val & mask) == mask;
+}
+
+static void _delay(unsigned us)
+{
+    if (IS_USED(MODULE_ZTIMER_USEC)) {
+        ztimer_sleep(ZTIMER_USEC, us);
+    } else if (IS_USED(MODULE_ZTIMER_MSEC)) {
+        ztimer_sleep(ZTIMER_MSEC, 1);
+    } else {
+        volatile unsigned count = (us * CLOCK_CORECLOCK) / US_PER_SEC;
+        while (--count) {}
```

Comment:
This could profit from a small comment. Also, what would happen if the CPU is taking significantly less than 1 µs per loop iteration? Or specifically, if the delay ends up only being a small fraction of what is requested, is there any issue to expect?

I wouldn't see to much issue in this just depending on `ztimer_msec`. There are few use cases that require a sd card reader but don't use a software timer anyway, and even fewer that cannot accept the additional ROM `ztimer_msec` adds.

Reply:
The idea was to being able to use this with [#17379](https://github.com/RIOT-OS/RIOT/pull/17379) (but I did not test this yet).

This takes core clock into account, but assumes the CPU won't do more than one loop iteration per cycle.